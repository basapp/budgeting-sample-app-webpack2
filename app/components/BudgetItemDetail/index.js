import * as React from 'react';
import { formatPercent } from 'utils/formatAmount';
import type { Transaction } from 'modules/transactions';
import styles from './style.scss';
import DonutChart from 'components/DonutChart';

type BudgetItemDetailProps = {
    transaction: Transaction,
    category: {},
    percent: number,
    redirectToBudget: {},
};

const categoryDataItem = (categoryId, value, description) => {
    return {
        categoryId,
        value,
        category: description,
    };
}
const BudgetItemDetail = ({ transaction, category, percent, redirectToBudget }: BudgetItemDetailProps) => {

    const isValidTransaction = transaction.id != undefined;
    const invalidTransactionMessage = "Transaction not found";
    const { id, categoryId, description } = transaction;
    const formatedPercent = formatPercent(percent);
    const isNegative = transaction.value < 0;
    const amountCls = isNegative ? styles.neg : styles.pos;

    const data = [
        categoryDataItem(categoryId, percent, description),
        categoryDataItem(0, 100 - percent, "Others"),
    ];

    const BackButton = () => {
        return <input type="submit"
            value="Back"
            className={styles.button}
            onClick={redirectToBudget} />
    };
    return (

        <div className={styles.budgetItem}>
            {isValidTransaction ? (<div>
                <h2>{description} <span className={styles.category}>{category}</span></h2>
                <h4><span className={amountCls}>{formatedPercent}</span></h4>
                <DonutChart data={data} dataLabel="category" dataKey="categoryId" usePercentSign={true} />
                <BackButton />
            </div>
            ) : (<div>
                <h2><span className={styles.notFound}>{invalidTransactionMessage}</span></h2><br/>
                <BackButton />
            </div>)}
        </div>
    );
};

export default BudgetItemDetail;