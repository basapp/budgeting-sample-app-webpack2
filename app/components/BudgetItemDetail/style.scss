@import 'theme/variables';

.budgetItem {
  display: flex;
  flex-flow: column;  
  font-size: 0.8em;  
  @media only screen and (max-width: $screen-small) {    
    align-items: center;   
    text-align: center;    
  }    
}

.neg {
  color: $red;
}
.neg:before {
  content: "-";
}
.pos {
  color: $green;
}
.pos:before {
  content: "+";
}
.category {
  font-weight: lighter;
  font-style: italic;
  color:rgb(72, 72, 72); 
}
.category:after {
  content: "]";
}
.category:before {
  content: "[";
}
.button {
  color: #fff;  
  text-transform: uppercase;
  background: rgb(28, 184, 65);
  text-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: .5em 1em;
  text-decoration: none;
  border: transparent;

  width: 60px;
  @media only screen and (max-width: $screen-small) {    
    width:100%;
  }  
}

.notFound {
  color: $red;  
}