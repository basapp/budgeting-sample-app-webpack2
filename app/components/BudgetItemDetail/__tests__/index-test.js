import React from 'react';
import renderer from 'react-test-renderer';
import BudgetItemDetail from 'components/BudgetItemDetail';

// mock nested components
jest.mock('components/DonutChart/Path');
jest.mock('components/Chart', () => 'div');
jest.mock('components/Legend', () => 'div');

it('renders correctly', () => {
  const mockTransaction = {
    id: 1,
    description: "Trader Joe's food",
    value: -423.34,
    categoryId: 1,
  };

  const mockCategory = 'Groceries';

  const mockPercent = 9.23;

  const tree = renderer.create(    
      <BudgetItemDetail transaction={mockTransaction} category={mockCategory} percent={mockPercent} redirectToBudget={()=>{}} />          
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly no Transaction', () => {
  const mockTransaction = {};

  const mockCategory = '';

  const mockPercent = 0;

  const tree = renderer.create(    
      <BudgetItemDetail transaction={mockTransaction} category={mockCategory} percent={mockPercent} redirectToBudget={()=>{}} />          
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
