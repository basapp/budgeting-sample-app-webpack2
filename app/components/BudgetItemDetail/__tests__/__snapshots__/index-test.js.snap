// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="budgetItem"
>
  <div>
    <h2>
      Trader Joe's food
       
      <span
        className="category"
      >
        Groceries
      </span>
    </h2>
    <h4>
      <span
        className="neg"
      >
        %9.23
      </span>
    </h4>
    <div
      className="donutChart"
    >
      <div
        height={316}
        padding={8}
        transform="translate(150,150)"
        width={316}
      />
      <div
        color={[Function]}
        data={
          Array [
            Object {
              "category": "Trader Joe's food",
              "categoryId": 1,
              "value": 9.23,
            },
            Object {
              "category": "Others",
              "categoryId": 0,
              "value": 90.77,
            },
          ]
        }
        dataKey="categoryId"
        dataLabel="category"
        dataValue="value"
        usePercentSign={true}
      />
    </div>
    <input
      className="button"
      onClick={[Function]}
      type="submit"
      value="Back"
    />
  </div>
</div>
`;

exports[`renders correctly no Transaction 1`] = `
<div
  className="budgetItem"
>
  <div>
    <h2>
      <span
        className="notFound"
      >
        Transaction not found
      </span>
    </h2>
    <br />
    <input
      className="button"
      onClick={[Function]}
      type="submit"
      value="Back"
    />
  </div>
</div>
`;
